plugins/modules/ec2_asg_instance_refresh.py

```diff
@@ -0,0 +1,190 @@
+description:
+  - Start or cancel an ec2 Auto Scaling Group Instance Refresh in AWS
+  - Can be used with ec2_asg_instance_refreshes_info to track the subsequent progress
+requirements: [ boto3 ]
+author: "Dan Khersonsky (@danquixote)"
+options:
+  state:
+    description:
+      - Desired state
+    type: str
+    required: true
+    choices: [ 'started', 'canceled' ]
+  name:
+    description:
+      - The name of the auto scaling group you are searching for.
+    type: str
+    required: true
+  strategy:
+    description:
+      - list of InstanceRefreshIds
+    type: str
+    default: 'Rolling'
+  preferences:
+    description:
+      - preferences
+    required: false
+    suboptions:
+      min_healthy_percentage:
+        description:
+          - min_healthy_percentage
+        type: int
+      instance_warmup:
+        description:
+          - instance_warmup
+        type: int
+    type: dict
+extends_documentation_fragment:
+- amazon.aws.aws
+- amazon.aws.ec2
+
+'''
+
+EXAMPLES = '''
+# Note: These examples do not set authentication details, see the AWS Guide for details.
+
+- name: Start a refresh
+  community.aws.ec2_asg_instance_refresh:
+    name: some-asg
+    state: started
+
+- name: Cancel a refresh
+  community.aws.ec2_asg_instance_refresh:
+    name: some-asg
+    state: canceled
+
+- name: Start a refresh and pass preferences
+  community.aws.ec2_asg_instance_refresh:
+    name: some-asg
+    state: started
+    preferences:
+      min_healthy_percentage: 91
+      instance_warmup: 60
+
+'''
+
+RETURN = '''
+---
+instance_refresh_id:
+    description: instance refresh id
+    returned: success
+    type: str
+    sample: "08b91cf7-8fa6-48af-b6a6-d227f40f1b9b"
+'''
+
+import re
+import sys
+
+try:
+    from botocore.exceptions import BotoCoreError, ClientError
+except ImportError:
+    pass  # caught by AnsibleAWSModule
+
+from ansible_collections.amazon.aws.plugins.module_utils.core import AnsibleAWSModule
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import camel_dict_to_snake_dict
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import AWSRetry
+
+
+def start_or_cancel_instance_refresh(conn, module):
+    """
+    Args:
+        conn (boto3.AutoScaling.Client): Valid Boto3 ASG client.
+        name (str): Mandatory name of the ASG you are looking for.
+        state (str): Start or Cancel a refresh
+
+    Returns:
+        Dict
+        {
+            'instance_refresh_id': 'string'
+        }
+    """
+
+    asg_state = module.params.get('state')
+    asg_name = module.params.get('name')
+    preferences = module.params.get('preferences')
+
+    args = {}
+    args['AutoScalingGroupName'] = asg_name
+    if asg_state == 'started':
+        args['Strategy'] = 'Rolling'
+    if preferences:
+        if asg_state == 'canceled':
+            module.fail_json_aws(Exception, msg='can not pass preferences dict when canceling a refresh')
```

Comment:
```suggestion
            module.fail_json(msg='can not pass preferences dict when canceling a refresh')
```

fail_json_aws is only needed when you've caught an exception.  It knows how to pull the relevant information out of BotoCoreError and ClientError

Reply:
Fixed in 42cf12ea1c666160e310eb94412e4e0898c77876